File: src/main/java/de/hglabor/plugins/training/challenges/mlg/streaks/data/StreakData.java
package de.hglabor.plugins.training.challenges.mlg.streaks.data;

import de.hglabor.plugins.training.challenges.mlg.streaks.StreakPlayer;
import org.bukkit.util.io.BukkitObjectInputStream;
import org.bukkit.util.io.BukkitObjectOutputStream;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.Serializable;
import java.util.HashMap;
import java.util.UUID;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

public class StreakData implements Serializable {
    private static transient final long serialVersionUID = -1681012206529286330L;
    public final HashMap<UUID, StreakPlayer> streakPlayers;
    public StreakData(HashMap<UUID, StreakPlayer> streakPlayers) {
        this.streakPlayers = streakPlayers;
    }
    public StreakData(StreakData loadedStreakData) {
        this.streakPlayers = loadedStreakData.streakPlayers;
    }

    public boolean saveData(String filePath) {
        try {
            FileOutputStream fileOut = new FileOutputStream(filePath);
            GZIPOutputStream gzOut = new GZIPOutputStream(fileOut);
            BukkitObjectOutputStream out = new BukkitObjectOutputStream(gzOut);
            out.writeObject(this);
            out.close();
            return true;
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        }
    }

    public static StreakData loadData(String filePath) {
        try {
            BukkitObjectInputStream in = new BukkitObjectInputStream(new GZIPInputStream(new FileInputStream(filePath)));
            StreakData streakData = (StreakData) in.readObject();
            in.close();
            return streakData;
        } catch (ClassNotFoundException | IOException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: src/main/kotlin/de/hglabor/plugins/training/data/objects/StreakData.kt
package de.hglabor.plugins.training.data.objects

import de.hglabor.plugins.training.challenges.mlg.streaks.StreakPlayer
import de.hglabor.plugins.training.challenges.mlg.streaks.StreakPlayers
import de.hglabor.plugins.training.data.loadObj
import de.hglabor.plugins.training.data.saveObj
import java.io.Serializable
import java.util.*

class StreakData() : Serializable {
    var streakPlayers: HashMap<UUID, StreakPlayer> = StreakPlayers.getStreakPlayers()

    companion object {
        const val serialVersionUID: Long = -6867189741698444290

        private const val PATH = "streaks"
        fun load() {
            loadObj(PATH)?.let {
                StreakPlayers.setStreakPlayers(StreakData(it as StreakData).streakPlayers)
            }
        }

        fun save() {
            saveObj(PATH, StreakData())
        }
    }

    // For loading
    constructor(data: StreakData): this() {
        this.streakPlayers = data.streakPlayers
    }
}